ouroboros-consensus/test/consensus-test/Test/Consensus/MiniProtocol/ObjectDiffusion/Smoke.hs

```diff
@@ -0,0 +1,302 @@
+  )
+import Ouroboros.Network.ControlMessage (ControlMessage (..))
+import Ouroboros.Network.NodeToNode.Version (NodeToNodeVersion (..))
+import Ouroboros.Network.Protocol.ObjectDiffusion.Codec (codecObjectDiffusionId)
+import Ouroboros.Network.Protocol.ObjectDiffusion.Inbound
+  ( ObjectDiffusionInboundPipelined
+  , objectDiffusionInboundPeerPipelined
+  )
+import Ouroboros.Network.Protocol.ObjectDiffusion.Outbound
+  ( ObjectDiffusionOutbound
+  , objectDiffusionOutboundPeer
+  )
+import Ouroboros.Network.Protocol.ObjectDiffusion.Type
+  ( NumObjectIdsReq (..)
+  , NumObjectsOutstanding (..)
+  , NumObjectsReq (..)
+  , ObjectDiffusion
+  )
+import Test.QuickCheck
+import Test.Tasty
+import Test.Tasty.QuickCheck
+import Test.Util.Orphans.Arbitrary ()
+import Test.Util.Orphans.IOLike ()
+
+tests :: TestTree
+tests =
+  testGroup
+    "ObjectDiffusion.Smoke"
+    [ testProperty
+        "ObjectDiffusion smoke test with mock objects"
+        prop_smoke
+    ]
+
+{-------------------------------------------------------------------------------
+  Provides a way to generate lists composed of objects with no duplicate ids,
+  with an Arbitrary instance
+-------------------------------------------------------------------------------}
+
+class WithId a idTy | a -> idTy where
+  getId :: a -> idTy
+
+newtype ListWithUniqueIds a idTy = ListWithUniqueIds [a]
+  deriving (Eq, Show, Ord)
+
+instance (Ord idTy, WithId a idTy, Arbitrary a) => Arbitrary (ListWithUniqueIds a idTy) where
+  arbitrary = ListWithUniqueIds . nubOrdOn getId <$> arbitrary
+
+instance WithId SmokeObject SmokeObjectId where getId = getSmokeObjectId
+
+{-------------------------------------------------------------------------------
+  Mock objectPools
+-------------------------------------------------------------------------------}
+
+newtype SmokeObjectId = SmokeObjectId Int
+  deriving (Eq, Ord, Show, NoThunks, Arbitrary)
+
+newtype SmokeObject = SmokeObject {getSmokeObjectId :: SmokeObjectId}
+  deriving (Eq, Ord, Show, NoThunks, Arbitrary)
+
+newtype SmokeObjectPool m = SmokeObjectPool (StrictTVar m [SmokeObject])
+
+newObjectPool :: MonadSTM m => [SmokeObject] -> m (SmokeObjectPool m)
+newObjectPool initialPoolContent = SmokeObjectPool <$> uncheckedNewTVarM initialPoolContent
+
+makeObjectPoolReader ::
+  MonadSTM m => SmokeObjectPool m -> ObjectPoolReader SmokeObjectId SmokeObject Int m
+makeObjectPoolReader (SmokeObjectPool poolContentTvar) =
+  ObjectPoolReader
+    { oprObjectId = getSmokeObjectId
+    , oprObjectsAfter = \minTicketNo limit -> do
+        poolContent <- readTVar poolContentTvar
+        pure $
+          take (fromIntegral limit) $
+            drop (minTicketNo + 1) $
+              ( (\(ticketNo, smokeObject) -> (ticketNo, getSmokeObjectId smokeObject, pure smokeObject))
+                  <$> zip [(0 :: Int) ..] poolContent
+              )
+    , oprZeroTicketNo = -1 -- objectPoolObjectIdsAfter uses strict comparison, and first ticketNo is 0.
```

Comment:
This is a bit confusing. Probably users won't see this -1, but still. I wonder if it's possible to either start numbering tickets at 1, or use a different name for `oprZeroTicketNo`. Also it's probably not a big thing, but we're losing half of the space by using -1 as the number before the first ticket number.